lib/src/block_parser.dart

```diff
@@ -728,7 +730,137 @@ class ParagraphSyntax extends BlockSyntax {
       parser.advance();
     }
 
-    var contents = parser.document.parseInline(childLines.join('\n'));
-    return new Element('p', contents);
+    var paragraphLines = _extractReflinkDefinitions(parser, childLines);
+    if (paragraphLines == null) {
+      // Paragraph consisted solely of reference link definitions.
+      return new Text('');
+    } else {
+      var contents = new UnparsedContent(paragraphLines.join('\n'));
+      return new Element('p', [contents]);
+    }
+  }
+
+  // Extract reference link definitions from the front of the paragraph, and
+  // return the remaining paragraph lines.
+  List<String> _extractReflinkDefinitions(
+      BlockParser parser, List<String> lines) {
```

Comment:
Yeah, it does seem pretty complex. However, if I'm reading the spec right, I still think there should be a more straightforward solution. As far as I can tell, parsing it doesn't require complex backtracking. My understanding is that, given a list of lines, you are trying to split it into:

1. The contiguous list of link definitions, each of which may be more than one line.
2. All of the remaining lines.

So, for any given list of lines, the links always come first, and there are no gaps between them. So I think parsing it could be something roughly like:

```dart
/// Parses a ref link starting at [start] in [lines]. Returns the number of
/// lines consumed or 0 if a ref link does not start there.
int parseRefLink(List<String> lines, int start) {
  // Basically the same thing as _parseReflinkDefinition(), but return the
  // number of parsed lines or zero on failure...
}

List<String> _extractReflinkDefinitions(
    BlockParser parser, List<String> lines) {
  var start = 0;
  while (true) {
    var consumed = parseRefLink(lines);
    if (consumed == 0) break;
    start += consumed;
  }

  // The rest of the lines are not links.
  return lines.getRange(start);
}
```

Does that work?